tests/unit/s2n_async_pkcs11_sign_test.c

```diff
@@ -0,0 +1,574 @@
+
+    POSIX_GUARD(s2n_cert_chain_and_key_set_ctx(rsa_chain_and_key, pkcs11_rsa_key_label));
+    POSIX_GUARD(s2n_cert_chain_and_key_set_ctx(ecdsa_chain_and_key, pkcs11_ec_key_label));
+
+    struct s2n_cipher_suite *test_cipher_suites[] = {
+        &s2n_rsa_with_aes_128_gcm_sha256,
+        &s2n_ecdhe_rsa_with_aes_128_gcm_sha256,
+        &s2n_ecdhe_ecdsa_with_aes_128_cbc_sha256,
+        &s2n_ecdhe_ecdsa_with_aes_256_cbc_sha384,
+    };
+
+    for(int i=0; i < sizeof(test_cipher_suites)/sizeof(test_cipher_suites[0]); i++) {
+        struct s2n_cipher_preferences server_cipher_preferences = {
+            .count = 1,
+            .suites = &test_cipher_suites[i],
+        };
+
+        struct s2n_security_policy server_security_policy = {
+            .minimum_protocol_version = S2N_TLS12,
+            .cipher_preferences = &server_cipher_preferences,
+            .kem_preferences = &kem_preferences_null,
+            .signature_preferences = &s2n_signature_preferences_20200207,
+            .ecc_preferences = &s2n_ecc_preferences_20200310,
+        };
+
+        EXPECT_TRUE(test_cipher_suites[i]->available);
+
+        TEST_DEBUG_PRINT("Testing %s\n", test_cipher_suites[i]->name);
+
+        /*  Test: RSA/ECDSA */
+        {
+            struct s2n_config *server_config, *client_config;
+            EXPECT_NOT_NULL(server_config = s2n_config_new());
+            EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(server_config, rsa_chain_and_key));
+            EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(server_config, ecdsa_chain_and_key));
+            EXPECT_SUCCESS(s2n_config_add_dhparams(server_config, dhparams_pem));
+            EXPECT_SUCCESS(s2n_config_set_async_pkey_callback(server_config, async_pkey_callback));
+            EXPECT_SUCCESS(s2n_config_set_verify_host_callback(server_config, verify_host_fn, NULL));
+            server_config->security_policy = &server_security_policy;
+
+            EXPECT_NOT_NULL(client_config = s2n_config_new());
+
+            /* Second half of suites is ECDSA. */
+            if(i >= 2) {
+                EXPECT_SUCCESS(s2n_config_set_verification_ca_location(server_config, S2N_ECDSA_P256_PKCS1_CERT_CHAIN, NULL));
+
+                EXPECT_SUCCESS(s2n_config_set_cipher_preferences(client_config, "20190214"));
+                EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(client_config, ecdsa_chain_and_key));
+                EXPECT_SUCCESS(s2n_config_set_verification_ca_location(client_config, S2N_ECDSA_P256_PKCS1_CERT_CHAIN, NULL));
+            } else {
+                EXPECT_SUCCESS(s2n_config_set_verification_ca_location(server_config, S2N_DEFAULT_TEST_CERT_CHAIN, NULL));
+
+                EXPECT_SUCCESS(s2n_config_set_verification_ca_location(client_config, S2N_DEFAULT_TEST_CERT_CHAIN, NULL));
+                EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(client_config, rsa_chain_and_key));
+            }
```

Comment:
You can probably get rid of all of this if you turn off certificate verification.